doc/specs/stdlib_str2num.md

```diff
@@ -0,0 +1,105 @@
+---
+title: str2num
+---
+
+# The `stdlib_str2num` module
+
+This module proposes a function-style interface for string-to-number conversion. It also profits from Fortran's interfaces to implement precision-dependant algorithms to maximize runtime efficiency.
+
+[TOC]
+
+## `to_num` - conversion of strings to numbers
+
+### Status
+
+Experimental
+
+### Description
+
+Convert a string or an array of strings to numerical types.
+
+### Syntax
+
+`number = [[stdlib_str2num(module):to_num(interface)]](string, mold)`
+
+### Arguments
+
+`string`: argument has `intent(in)` and is of type `character(*)`.
+
+`mold`: argument has `intent(in)` and is of numerical type. currently: `integer`, `real32`, `real64` or `real128`. **Note**: the mold argument is included to help compilers chose the correct implementation at compile-time. Currently, compilers are not able to disambiguate functions with respect to the left-hand-side of an assignment.
+
+### Return value
+
+Return a scalar of numerical type (`integer`, `real32` or `real64`).
+
+### Example
+
+```fortran
+program example_string_to_number
+  use stdlib_kinds, only: dp
+  use stdlib_str2num, only: to_num
+  implicit none
+  character(:), allocatable :: txt
+  real(dp) :: x
+
+  txt = ' 8.8541878128e−12 '
+  x = to_num( txt , x )
+end program example_string_to_number
+```
+
+## `to_num_p` - conversion of a stream of values in a string to numbers
+
+### Status
+
+Experimental
+
+### Description
+
+Convert a stream of values in a string to an array of values.
+
+### Syntax
+
+`number = [[stdlib_str2num(module):to_num_p(interface)]](string, mold)`
+
+### Arguments
+
+`string`: argument has `intent(in)` and is of type `character(*), pointer`.
+
+`mold`: argument has `intent(in)` and is of numerical type. currently: `integer`, `real32`, `real64` or `real128`. **Note**: the mold argument is included to help compilers chose the correct implementation at compile-time. Currently, compilers are not able to disambiguate functions with respect to the left-hand-side of an assignment.
+
+### Return value
+
+Return a scalar of numerical type (`integer`, `real32` or `real64`).
```

Comment:
```suggestion
Return a scalar of numerical type (`int8`, int16`, int32`, int64`, `real32`, `real64` or `real128`).
```
I wonder if we should use e.g., `sp` instead of `real32`.

Reply:
Wonder the same thing. For the doc I think the most explicit would be the best.

Reply:
See comment below: Shortly, the following could be written (similarly to other specs):
```suggestion
Return a scalar of numerical type (i.e., `integer` or `real`).
```

this will be not fypp/compiler-dependent